x-pack/plugins/lists/server/services/items/update_list_item.ts

```diff
@@ -53,30 +57,72 @@ export const updateListItem = async ({
     if (elasticQuery == null) {
       return null;
     } else {
-      const doc: UpdateEsListItemSchema = {
+      checkVersionConflict(_version, listItem._version);
```

Comment:
We have a gap here between the fetching/checking the version and and executing the updateByQuery so we can't be sure that other simultaneous updates aren't being overwritten. What do you think about switching the version check to be part of the updateByQuery script using [ctx['_version']](https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-update-by-query-context.html)? I'm thinking something like setting `ctx.op='noop'` (https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html#docs-update-by-query-api-source) if the versions don't match when we check as part of the script.

We'd have to switch from using the encoded version string to the doc version, but theoretically it's not a breaking change since the version we return in the APIs is an opaque string.

Reply:
> We have a gap here between the fetching/checking the version and and executing the updateByQuery so we can't be sure that other simultaneous updates aren't being overwritten. What do you think about switching the version check to be part of the updateByQuery script 

so, _version would present somehow limited ability to control history of updates, as it does not account multiple shards. On the other hand, lists/items configured as single shard when created. So, users would need to manually changed system index template to add more shards? Ot there any other way, these indices can be created with multiple shards?


> We'd have to switch from using the encoded version string to the doc version, but theoretically it's not a breaking change since the version we return in the APIs is an opaque string.

we have it [documented](https://www.elastic.co/guide/en/security/current/lists-api-update-item.html#_request_body_19) as base64 encoding of if_seq_no and if_primary_term 
Is this still an option for us to move to _version and don't account for multiple shards OCC?

Reply:
or maybe we can get internal version from getListItem, and use it inside painless script?
So we will do 2 checks:
1. current one, with OCC
2. in query, using _version, which 'passed' conflict validation
In this way we won't need to change external API or documentation

Reply:
> _version would present somehow limited ability to control history of updates, as it does not account multiple shards

Can you expand on this? What problems does _version have with multiple shards?

Reply:
From my understanding, only reliable OCC is supported by [if_seq_no and if_primary_term ](https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html)

Link from https://discuss.elastic.co/t/relation-between-version-seq-no-and-primary-term/179647 leads to https://github.com/elastic/elasticsearch/issues/19269#issuecomment-488598561, where it mentioned
> the _version field continues not to uniquely identify a particular version of a document; if you need to do this then you should move to using the _seq_no and _primary_term fields.

[Here](https://www.elastic.co/guide/en/elasticsearch/resiliency/current/index.html) is given a scenario, when _version is not enough

> When a primary has been partitioned away from the cluster there is a short period of time until it detects this. During that time it will continue indexing writes locally, thereby updating document versions. When it tries to replicate the operation, however, it will discover that it is partitioned away. It won’t acknowledge the write and will wait until the partition is resolved to negotiate with the master on how to proceed. The master will decide to either fail any replicas which failed to index the operations on the primary or tell the primary that it has to step down because a new primary has been chosen in the meantime. Since the old primary has already written documents, clients may already have read from the old primary before it shuts itself down. The _version field of these reads may not uniquely identify the document’s version if the new primary has already accepted writes for the same document (see [#19269](https://github.com/elastic/elasticsearch/issues/19269)).
> The Sequence numbers infrastructure [#10708](https://github.com/elastic/elasticsearch/issues/10708) has introduced more precise ways for tracking primary changes. This new infrastructure therefore provides a way for uniquely identifying documents using their primary term and sequence number fields, even in the presence of network partitions, and has been used to replace the _version field in operations that require uniquely identifying the document, such as optimistic concurrency control.